derp/derphttp/derphttp_client.go

```diff
@@ -56,6 +56,11 @@ type Client struct {
 	MeshKey   string             // optional; for trusted clients
 	IsProber  bool               // optional; for probers to optional declare themselves as such
 
+	// BaseContext, if non-nil, returns the base context to use for dialing a
```

Comment:
Non-blocking idea, maybe out of scope here.

One thing I dislike about the http.Server option is that it doesn't let you get hold of the CancelFunc, if the context has one.

What do you think about maybe:

```go
BaseContext func() (context.Context, context.CancelFunc)
```

and then have newContext just return `func() {}` along with background?

Reply:
Why does derphttp want the CancelFunc?

The goal of this is entirely context values, not deadlines/timeouts.

Also: it's sounding like Go 1.22 will have no cost for un-canceled timers and they'll just be GC'ed away whenever possible.

Reply:
Ah, I was thinking of default dial timeouts. But that's cool about the runtime handling it better soon!